File: src/android/java/app/polarbear/KeyboardAccessibilityService.java
package app.polarbear;

import android.accessibilityservice.AccessibilityService;
import android.accessibilityservice.AccessibilityServiceInfo;
import android.util.Log;
import android.view.KeyEvent;
import android.view.accessibility.AccessibilityEvent;

public class KeyboardAccessibilityService extends AccessibilityService {
    private static final String TAG = "LocalDesktopA11y";

    static {
        System.loadLibrary("localdesktop");
    }

    @Override
    protected void onServiceConnected() {
        super.onServiceConnected();

        Log.i(TAG, "Keyboard accessibility service connected");
        nativeSetServiceConnected(true);

        AccessibilityServiceInfo info = new AccessibilityServiceInfo();
        info.packageNames = new String[] {getPackageName()};
        info.eventTypes = AccessibilityEvent.TYPE_WINDOW_STATE_CHANGED;
        info.notificationTimeout = 50;
        info.flags = AccessibilityServiceInfo.FLAG_REQUEST_FILTER_KEY_EVENTS;
        info.feedbackType = AccessibilityServiceInfo.FEEDBACK_GENERIC;
        setServiceInfo(info);
    }

    @Override
    protected boolean onKeyEvent(KeyEvent event) {
        return nativeOnKeyEvent(
                event.getAction(),
                event.getKeyCode(),
                event.getScanCode(),
                event.getEventTime())
            || super.onKeyEvent(event);
    }

    @Override
    public void onAccessibilityEvent(AccessibilityEvent event) {}

    @Override
    public void onInterrupt() {}

    @Override
    public void onDestroy() {
        nativeSetServiceConnected(false);
        super.onDestroy();
    }

    private static native void nativeSetServiceConnected(boolean connected);

    private static native boolean nativeOnKeyEvent(
        int action,
        int keyCode,
        int scanCode,
        long eventTime
    );
}

File: src/android/kotlin/KeyboardAccessibilityService.kt
package app.polarbear

import android.accessibilityservice.AccessibilityService
import android.accessibilityservice.AccessibilityServiceInfo
import android.util.Log
import android.view.KeyEvent
import android.view.accessibility.AccessibilityEvent

class KeyboardAccessibilityService : AccessibilityService() {
    override fun onServiceConnected() {
        super.onServiceConnected()

        Log.i(TAG, "Keyboard accessibility service connected")
        nativeSetServiceConnected(true)

        serviceInfo = AccessibilityServiceInfo().apply {
            packageNames = arrayOf(packageName)
            eventTypes = AccessibilityEvent.TYPES_ALL_MASK
            notificationTimeout = 100
            flags = AccessibilityServiceInfo.FLAG_REQUEST_FILTER_KEY_EVENTS
            feedbackType = AccessibilityServiceInfo.FEEDBACK_SPOKEN
        }
    }

    override fun onKeyEvent(event: KeyEvent): Boolean {
        return nativeOnKeyEvent(
            event.action,
            event.keyCode,
            event.scanCode,
            event.eventTime,
        ) || super.onKeyEvent(event)
    }

    override fun onAccessibilityEvent(event: AccessibilityEvent) {
    }

    override fun onInterrupt() {
    }

    override fun onDestroy() {
        nativeSetServiceConnected(false)
        super.onDestroy()
    }

    private external fun nativeSetServiceConnected(connected: Boolean)
    private external fun nativeOnKeyEvent(
        action: Int,
        keyCode: Int,
        scanCode: Int,
        eventTime: Long,
    ): Boolean

    companion object {
        private const val TAG = "LocalDesktopA11y"

        init {
            System.loadLibrary("localdesktop")
        }
    }
}
